klippy/extras/pwm_tool.py

```diff
@@ -0,0 +1,187 @@
+# Queued PWM gpio output
+#
+# Copyright (C) 2017-2023  Kevin O'Connor <kevin@koconnor.net>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import chelper
+
+PIN_MIN_TIME = 0.100
+RESEND_HOST_TIME = 0.300 + PIN_MIN_TIME
+MAX_SCHEDULE_TIME = 5.0
+
+class error(Exception):
+    pass
+
+class MCU_queued_pwm:
+    def __init__(self, pin_params):
+        self._mcu = pin_params['chip']
+        self._hardware_pwm = False
+        self._cycle_time = 0.100
+        self._max_duration = 2.
+        self._oid = self._mcu.create_oid()
+        ffi_main, ffi_lib = chelper.get_ffi()
+        self._stepqueue = ffi_main.gc(ffi_lib.stepcompress_alloc(self._oid),
+                                      ffi_lib.stepcompress_free)
+        self._mcu.register_stepqueue(self._stepqueue)
```

Comment:
Minor cosmetic: It is not really a *step*-queue.

In object-oriented C++ I would have suggested a general buffer management queue (e.g. `command_queue`), which the stepcompress system would inherit the queue from and add the compression part (i.e. step-dir-filter etc.)

Perhaps I am thinking too general, but the reason for this is that this might be a possible location for a hypothetical *proportionality* feature (i.e. *S*-Parameter in [G0 in RepRap documentation](https://reprap.org/wiki/G-code#G-commands)).
After a stepcompress-flush, a toolhead might insert `pwm_tool` messages into the queue according to `actual speed /  requested speed` and a gamma-curve.

But we might "cross the bridge once we get there".

Reply:
Earlier this year I spent some time reworking the trapq, stepcompress, and steppersync code.  It's difficult to get all that code correct however, so I haven't gotten it to the point that it is ready to merge.  Largely, the pwm_tool code is largely orthogonal, so likely best not to introduce additional dependencies to it.